pkg/kademlia/pslice/pslice_test.go

```diff
@@ -0,0 +1,335 @@
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package pslice_test
+
+import (
+	"errors"
+	"testing"
+
+	"github.com/ethersphere/bee/pkg/kademlia/pslice"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"github.com/ethersphere/bee/pkg/topology/test"
+)
+
+// TestShallowestEmpty tests that ShallowestEmpty functionality works correctly.
+func TestShallowestEmpty(t *testing.T) {
+	var (
+		ps    = pslice.New(16)
+		base  = test.RandomAddress()
+		peers = make([]swarm.Address, 16)
+	)
+
+	for i := 0; i < 16; i++ {
+		a := test.RandomAddressAt(base, i)
+		peers[i] = a
+	}
+
+	for i, v := range peers {
+		ps.Add(v, uint8(i))
+		sd, none := ps.ShallowestEmpty()
+		if i == 15 {
+			if !none {
+				t.Fatal("expected last bin to be empty and return no empty bins")
+			}
+		} else {
+			if sd != uint8(i+1) {
+				t.Fatalf("expected shallow empty bin to be %d but got %d", i+1, sd)
+			}
+			if none {
+				t.Fatal("got no empty bins but wanted some")
+			}
+		}
+	}
+
+	for _, tc := range []struct {
+		rmPO  int
+		expSE uint8
+	}{
+		{rmPO: 3,
+			expSE: 3,
+		}, {
+			rmPO:  1,
+			expSE: 1,
+		}, {
+			rmPO:  10,
+			expSE: 1,
+		}, {
+			rmPO:  15,
+			expSE: 1,
+		}, {
+			rmPO:  14,
+			expSE: 1,
+		}, {
+			rmPO:  0,
+			expSE: 0,
+		},
+	} {
+		po := uint8(swarm.Proximity(base.Bytes(), peers[tc.rmPO].Bytes()))
+		ps.Remove(peers[tc.rmPO], po)
+
+		sd, none := ps.ShallowestEmpty()
+		if sd != tc.expSE || none {
+			t.Fatalf("empty bin mismatch got %d want %d", sd, tc.expSE)
+		}
+	}
+	ps.Add(peers[0], 0)
+	if sd, none := ps.ShallowestEmpty(); sd != 1 || none {
+		t.Fatalf("expected bin 1 to be empty shallowest but got %d", sd)
+	}
+}
+
+// TestAddRemove checks that the Add, Remove and Exists methods work as expected.
+func TestAddRemove(t *testing.T) {
+	var (
+		ps    = pslice.New(4)
+		base  = test.RandomAddress()
+		peers = make([]swarm.Address, 8)
+	)
+
+	// 2 peers per bin
+	// indexes {0,1} {2,3} {4,5} {6,7}
+	for i := 0; i < 8; i += 2 {
+		a := test.RandomAddressAt(base, i)
+		peers[i] = a
+
+		b := test.RandomAddressAt(base, i)
+		peers[i+1] = b
+	}
+	chkNotExists(t, ps, peers...)
+
+	// add one
+	ps.Add(peers[0], 0)
+	chkLen(t, ps, 1)
+	chkExists(t, ps, peers[:1]...)
+	chkNotExists(t, ps, peers[1:]...)
+
+	// check duplicates
+	ps.Add(peers[0], 0)
+	chkLen(t, ps, 1)
+	chkBins(t, ps, []uint{0, 1, 1, 1})
+	chkExists(t, ps, peers[:1]...)
+	chkNotExists(t, ps, peers[1:]...)
+
+	// check empty
+	ps.Remove(peers[0], 0)
+	chkLen(t, ps, 0)
+	chkBins(t, ps, []uint{0, 0, 0, 0})
+	chkNotExists(t, ps, peers...)
+
+	// add two in bin 0
+	ps.Add(peers[0], 0)
+	ps.Add(peers[1], 0)
+	chkLen(t, ps, 2)
+	chkBins(t, ps, []uint{0, 2, 2, 2})
+	chkExists(t, ps, peers[:2]...)
+	chkNotExists(t, ps, peers[2:]...)
+
+	ps.Add(peers[2], 1)
+	ps.Add(peers[3], 1)
+	chkLen(t, ps, 4)
+	chkBins(t, ps, []uint{0, 2, 4, 4})
+	chkExists(t, ps, peers[:4]...)
+	chkNotExists(t, ps, peers[4:]...)
```

Comment:
is it necessary to keep checking peers that cannot have been added?

Reply:
i dont see much harm. yes it might be a trivial check but i don't see it as harmful

Reply:
> a trivial check

I had to look twice when reading through the test to understand that it's not testing anything really. It does not help clarity.